### ISSUE TITLE
Interface casting issues when an unknown type is returned

### ISSUE THREAD

**Issue by cameronwhite** [gircore/gir.core#838]
I ran into an interesting scenario with the following code to try obtain a `Gio.File` from the `Gio.ListModel` returned by `Gtk.FileChooser.GetFiles()`

```csharp
Gtk.FileChooserDialog file_chooser = ... // open some files with a file chooser

Gio.ListModel files = file_chooser.GetFiles ();
for (uint i = 0, n = files.GetNItems (); i < n; ++i) {
	// The runtime type is just a GObject.Object instance
	// The gtype is 'GLocalFile' which is a private type that inherits from GObject and implements the 'GFile' interface
	GObject.Object? obj = files.GetObject (i);
        var file = obj as Gio.File; // fails
}
```

The GObject returned is of type `GLocalFile` which isn't public, so there isn't a C# type to create an instance of. The `ObjectWrapper` then just creates an instance of the parent class, `GObject.Object` in this case.
This presents a problem with trying to cast to a `Gio.File` interface, since the C# object doesn't have a runtime type that implements that interface

Perhaps some sort of cast function that checks the gtype is needed?
My manual workaround was to do `var file = new Gio.FileHelper (files.GetObject (i).Handle, ownedRef: true)`, although that doesn't do any safe type checking

**Comment by badcel** [comment]
For classes which implement interfaces inside list models this should work already correctly. For private classes which implement interfaces not.

Your suggestion using a method to cast the list items we could return an `IEnumarable<T>` which either casts a class to a more concrete class/interface or wraps the instance into an interface wrapper and checks the type of the gobject class is not known to us.

perhaps an additional method To get a casted single item would be helpful, too.

**Comment by cameronwhite** [comment]
Yeah, getting an `IEnumerable<T>` that does casting if necessary sounds like a good way to go. Even for cases that don't hit this interface casting issue, it would be a more C#-style API than having to use the `GetNItems()` to iterate over the list

**Comment by badcel** [comment]
#588 will help with integrating native types with some dotnet interfaces.

To return the correct type from a given pointer the type detection will be improved as part of #1118.
